src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -131,26 +131,50 @@ private void DeleteLeftoverDirectories()
 
         protected override string PreparePathToLoad(string originalAnalyzerPath, ImmutableHashSet<string> cultureNames)
         {
-            var analyzerFileName = Path.GetFileName(originalAnalyzerPath);
-            var shadowDirectory = CreateUniqueDirectoryForAssembly();
-            var shadowAnalyzerPath = Path.Combine(shadowDirectory, analyzerFileName);
-            copyFile(originalAnalyzerPath, shadowAnalyzerPath);
+            var mvid = ReadMvid(originalAnalyzerPath);
+            if (_mvidPathMap.TryGetValue(mvid, out Task<string>? copyTask))
+            {
+                return copyTask.Result;
+            }
 
-            if (cultureNames.IsEmpty)
+            var tcs = new TaskCompletionSource<string>();
+            var task = _mvidPathMap.GetOrAdd(mvid, tcs.Task);
+            if (object.ReferenceEquals(task, tcs.Task))
             {
+                // This thread won and we need to do the copy.
+                var shadowAnalyzerPath = copyAnalyzerContents();
+                tcs.SetResult(shadowAnalyzerPath);
                 return shadowAnalyzerPath;
```

Comment:
Consider wrapping this in a try/catch and fault the task if anything goes wrong. Otherwise those other things asking will just hang forever.

Reply:
Good call.